[JIT] Allow freezing modules that contain mutable interfaces

davidberard98 opened the pull request:
This PR allows freezing modules like the one below:
```python
# Ex. 1
        @torch.jit.interface
        class ModuleInterface(torch.nn.Module):
            def forward(self, inp: torch.Tensor) -> torch.Tensor:
                pass

        class ImplementsInterface(torch.nn.Module):
            def __init__(self):
                super(ImplementsInterface, self).__init__()
                self.sum = torch.zeros((2, 2))

            def forward(self, inp: torch.Tensor) -> torch.Tensor:
                self.sum += inp.relu()  # this makes the interface-implementing module mutable
                                        # and previously this would prevent freezing
                return self.sum

        class WrapperModule(torch.nn.Module):
            impl: ModuleInterface

            def __init__(self):
                super().__init__()
                self.impl = ImplementsInterface()

            def forward(self, x: torch.Tensor) -> torch.Tensor:
                return self.impl.forward(x)
```

Previously during freezing, we handle interfaces as shown below:
1. we inline interfaces in any preserved method graphs
2. during `cleanupFrozenModule`, we try to simplify the module data structure (<- this part is unrelated to freezing so far). During this step, if we found that a interface type was mutable, we'd error out; because of the possibility of a module that _swaps out the value of an interface-typed attribute at runtime_.

Below is an example of a module that swaps out the value of an interface-typed attribute at runtime:
```python
# Ex. 2
class MyBadModule(torch.nn.Module):
    impl: MyInterface
    option1: IfaceImpl1
    option2: IfaceImpl2
    ....
    def forward(self, x):
        if x > 0:
            self.impl = self.option1
        else:
            self.impl = self.option2
        ....
```

^ this type of situation cannot be supported by freezing (or at least would be difficult to do correctly) because it greatly complicates the details of handling types and simplifying the module data structure.

But we can still support the first example without _too_ much work:
1. inline the interface code as before
2. check to see if we have any setattrs on interface types; if so, error out
3. otherwise, replace the type of the interface types with the concrete type implementation
4. continue simplifying the module data structure as if we never had any interfaces.

davidberard98 committed:
add initial tests

davidberard98 committed:
add support for freezing mutable interfaces

davidberard98 committed:
comments

davidberard98 committed:
lint

davidberard98 committed:
remove expected failures

davidberard98 committed:
handle setattr check on non-module parent objects

eellison reviewed (commented):
looks good ! one question

davidberard98 commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -113,18 +126,23 @@ class AttributePropagator {
     };
 
+    std::unordered_map<std::string, std::unordered_set<std::string>>
+        interfacesToRetype;
```
I meant `interfacesToReassignType` haha... so yeah I guess I should rename that :)

davidberard98 commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -113,18 +126,23 @@ class AttributePropagator {
       }
       // Record Attributes that are explicitly set in the module.
       // They cannot be folded.
       recordMutableAttrs(graph);
```
@eellison suppose we put reassignInterfaceTypes immediately after inlineInterfaceCalls. Then we can have the following situation:

* preservedMethod1 does getattr(self, some_interface_module). It registers it as an interface to reassign type
* we reassign type for some_interface_module, so now the module object labels it as an interface type
* preservedMethod2 does setattr(self, some_interface_module). But when it looks it up in the module, it sees that some_interface_module is _not_ an interface (because the type was reassigned already). So it doesn't error out.

Is there anything that can go wrong with reassigning interface types after recordMutableAttrs?

eellison replied:
Yea, good point. Sorry the important part is that the calls themselves are inlined (and mutation is observed) not that the actual types on the modules are re-assigned.

davidberard98 replied:
Actually just to be safe I reordered it anyway, just added an extra loop to do the recordMutableAttrs afterwards.

eellison reviewed (approved):
Looks great!! Great tests too!!

davidberard98 committed:
address comments

- add tests (recursive interface swap & 2-method interface swap)
- reassign types _before_ checking for mutability

davidberard98 commented:
@pytorchbot rebase -s

davidberard98 commented:
@pytorchbot rebase -s

davidberard98 commented:
@pytorchbot merge
